File: src/main/java/waterpower/integration/ic2/IndustrialCraftModule.java
/**
 * Copyright (c) Huang Yuhui, 2014
 * 
 * "WaterCraft" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package waterpower.integration.ic2;

import ic2.api.recipe.RecipeInputItemStack;
import ic2.api.recipe.RecipeInputOreDict;
import ic2.api.recipe.Recipes;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import waterpower.common.block.ore.OreType;
import waterpower.common.item.GlobalItems;
import waterpower.common.item.crafting.ItemMaterial;
import waterpower.common.item.crafting.MaterialForms;
import waterpower.integration.BaseModule;

public class IndustrialCraftModule extends BaseModule {

	public static boolean IS_INDUSTRIAL_CRAFT_RECIPES_AVAILABLE = false;

	public static void compressor(ItemStack input, ItemStack output) {
		try {
			Recipes.compressor.addRecipe(new RecipeInputItemStack(input), null, false, output);
		} catch (Throwable t) {
			t.printStackTrace();
		}
	}

	public static void blastfurance(ItemStack input, ItemStack output) {
		try {
			Recipes.blastfurnace.addRecipe(new RecipeInputItemStack(input), null, false, output);
		} catch (Throwable t) {
			t.printStackTrace();
		}
	}

	public static void blastfurance(String input, ItemStack output) {
		try {
			Recipes.blastfurnace.addRecipe(new RecipeInputOreDict(input), null, false, output);
		} catch (Throwable t) {
			t.printStackTrace();
		}
	}

	public static void macerator(ItemStack input, ItemStack output) {
		try {
			Recipes.macerator.addRecipe(new RecipeInputItemStack(input), null, false, output);
		} catch (Throwable t) {
			t.printStackTrace();
		}
	}

	public static void metalformerRolling(ItemStack input, ItemStack output) {
		try {
			Recipes.metalformerRolling.addRecipe(new RecipeInputItemStack(input), null, false, output);
		} catch (Throwable t) {
			t.printStackTrace();
		}
	}

	public static void blockcutter(ItemStack input, ItemStack output) {
		try {
			Recipes.blockcutter.addRecipe(new RecipeInputItemStack(input), null, false, output);
		} catch (Throwable t) {
			t.printStackTrace();
		}
	}

	public static void oreWashing(ItemStack input, ItemStack... output) {
		try {
			NBTTagCompound metadata = new NBTTagCompound();
			metadata.setInteger("amount", 1000);
			Recipes.oreWashing.addRecipe(new RecipeInputItemStack(input), metadata, false, output);
		} catch (Throwable t) {
			t.printStackTrace();
		}
	}

	public static Object getMaceratorMachineManager() {
		try {
			return Recipes.macerator;
		} catch (Throwable t) {
			return null;
		}
	}

	public static Object getCompressorMachineManager() {
		try {
			return Recipes.compressor;
		} catch (Throwable t) {
			return null;
		}
	}

	public static Object getCutterMachineManager() {
		try {
			return Recipes.blockcutter;
		} catch (Throwable t) {
			return null;
		}
	}

	@Override
	public void postInit() {
		try {
			ItemStack item = ICItemFinder.getItem("dust", "stone"), iron = ICItemFinder.getItem("dust", "small_iron");

			for (OreType o : OreType.values())
				IndustrialCraftModule.oreWashing(new ItemStack(GlobalItems.oreDust, 1, o.ordinal()),
						ItemMaterial.get(o.t, MaterialForms.dust), iron, item);
		} catch (Throwable t) {
			t.printStackTrace();
		}
	}

}

File: src/main/kotlin/waterpower/integration/ic2/IndustrialCraftModule.kt
/**
 * Copyright (c) Huang Yuhui, 2017
 *
 * "WaterPower" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package waterpower.integration.ic2

import ic2.api.recipe.Recipes
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraftforge.fml.common.Optional
import waterpower.annotations.Integration
import waterpower.common.block.ore.Ores
import waterpower.common.init.WPItems
import waterpower.common.item.MaterialForms
import waterpower.integration.IDs
import waterpower.integration.IModule

@Integration(IDs.IndustrialCraft2)
object IndustrialCraftModule : IModule() {
    val RECIPE_AVAILABLE = true

    @Optional.Method(modid = IDs.IndustrialCraft2)
    fun compressor(input: ItemStack, output: ItemStack): Boolean {
        if (RECIPE_AVAILABLE)
            try {
                return Recipes.compressor.addRecipe(Recipes.inputFactory.forStack(input), null, false, output)
            } catch (t: Throwable) {
                t.printStackTrace()
            }
        return false
    }

    @Optional.Method(modid = IDs.IndustrialCraft2)
    fun blastFurnace(input: ItemStack, output: ItemStack, time: Int)
            = blastFurnace(input, output)

    @Optional.Method(modid = IDs.IndustrialCraft2)
    fun blastFurnace(input: ItemStack, output: ItemStack): Boolean {
        if (RECIPE_AVAILABLE)
            try {
                return Recipes.blastfurnace.addRecipe(Recipes.inputFactory.forStack(input), null, false, output)
            } catch (t: Throwable) {
                t.printStackTrace()
            }
        return false
    }

    @Optional.Method(modid = IDs.IndustrialCraft2)
    fun blastFurnace(input: String, output: ItemStack, time: Int): Boolean {
        if (RECIPE_AVAILABLE)
            try {
                return Recipes.blastfurnace.addRecipe(Recipes.inputFactory.forOreDict(input), null, false, output)
            } catch (t: Throwable) {
                t.printStackTrace()
            }
        return false
    }

    @Optional.Method(modid = IDs.IndustrialCraft2)
    fun macerator(input: ItemStack, output: ItemStack): Boolean {
        if (RECIPE_AVAILABLE)
            try {
                return Recipes.macerator.addRecipe(Recipes.inputFactory.forStack(input), null, false, output)
            } catch (t: Throwable) {
                t.printStackTrace()
            }
        return false
    }

    @Optional.Method(modid = IDs.IndustrialCraft2)
    fun metalformerRolling(input: ItemStack, output: ItemStack): Boolean {
        if (RECIPE_AVAILABLE)
            try {
                return Recipes.metalformerRolling.addRecipe(Recipes.inputFactory.forStack(input), null, false, output)
            } catch (t: Throwable) {
                t.printStackTrace()
            }
        return false
    }

    @Optional.Method(modid = IDs.IndustrialCraft2)
    fun blockcutter(input: ItemStack, output: ItemStack): Boolean {
        if (RECIPE_AVAILABLE)
            try {
                return Recipes.blockcutter.addRecipe(Recipes.inputFactory.forStack(input), null, false, output)
            } catch (t: Throwable) {
                t.printStackTrace()
            }
        return false
    }

    @Optional.Method(modid = IDs.IndustrialCraft2)
    fun oreWashing(input: ItemStack, vararg output: ItemStack): Boolean {
        if (RECIPE_AVAILABLE)
            try {
                val metadata = NBTTagCompound()
                metadata.setInteger("amount", 1000)
                return Recipes.oreWashing.addRecipe(Recipes.inputFactory.forStack(input), metadata, false, *output)
            } catch (t: Throwable) {
                t.printStackTrace()
            }
        return false
    }

    fun getMaceratorMachineManager(): Any? {
        try {
            return Recipes.macerator
        } catch (t: Throwable) {
            return null
        }
    }

    fun getCompressorMachineManager(): Any? {
        try {
            return Recipes.compressor
        } catch (t: Throwable) {
            return null
        }
    }

    fun getCutterMachineManager(): Any? {
        try {
            return Recipes.blockcutter
        } catch (t: Throwable) {
            return null
        }
    }

    override fun onPreInit() {
        waterpower.common.recipe.Recipes.benders += this::metalformerRolling
        waterpower.common.recipe.Recipes.blastFurnaces += this::blastFurnace
        waterpower.common.recipe.Recipes.blastFurnacesOreDict += this::blastFurnace
        waterpower.common.recipe.Recipes.crushers += this::macerator
        waterpower.common.recipe.Recipes.compressors += this::compressor
        waterpower.common.recipe.Recipes.cutters += this::blockcutter
    }

    override fun onPostInit() {
        try {
            val item = ICItemFinder.getItem("dust", "stone")
            val iron = ICItemFinder.getItem("dust", "small_iron")

            if (item == null || iron == null)
                return

            for (o in Ores.values())
                oreWashing(WPItems.crushed.getItemStack(o),
                        WPItems.material.getItemStack(o.material, MaterialForms.dust), iron, item)
        } catch (t: Throwable) {
            t.printStackTrace()
        }

    }
}